### PR TITLE
deps: Update protbuf to v23.1

### PR DESCRIPTION
**Pull request by yanavlasov**
New patches:

1. PGV to work with handling of new C++20 keywords added in protobuf (upstreaming change to PGV now)
2. absl patch to exclude stack symbolization for WASM modules. The implementation is buggy and causes existing WASM filter tests to fail. This is being fixed in emscripten SDK and absl https://github.com/emscripten-core/emscripten/issues/19753
3. Compilation issues in protobuf. Fixed on protbuf main, waiting for a new version to be cut

Breaking changes:

- Substitution formatter of the filter dynamic metadata no longer supports NaN and Inf floating point numbers, producing an error messages instead of JSON object of the proto. This is deemed to be an edge case.

Fixed a bunch of change detector tests that relied on specific error messages emitted by the previous version of the protobuf JSON parser. 

Some dependencies are not at the bleeding edge versions, but this PR brings Envoy code over the biggest hump and allows to upgrade to newer version with smaller changes later.

Risk Level: Low
Testing: Unit Tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A
Part of #27757

### CODE REVIEW

**Comment by phlax** [comment]
patch looks good, thanks for working on this!

**Comment by yanavlasov** [comment]
Slow progress. Many dependencies are broken. Actively working on it.

**Comment by phlax** [comment]
possibly related https://github.com/protocolbuffers/protobuf/issues/12523

**Comment by phlax** [comment]
re the `python_headers` im not sure we still need to alias them - but if we do i believe we can use 

```diff
diff --git a/bazel/repositories.bzl b/bazel/repositories.bzl
index fd13e73479..ac50ff6276 100644
--- a/bazel/repositories.bzl
+++ b/bazel/repositories.bzl
@@ -894,7 +894,7 @@ def _com_google_protobuf():
     # https://github.com/google/protobuf/blob/v3.6.1/util/python/BUILD#L6-L9
     native.bind(
         name = "python_headers",
-        actual = "@com_google_protobuf//util/python:python_headers",
+        actual = "@system_python//:python_headers",
     )
 
 def _io_opencensus_cpp():
```

**Comment by yanavlasov** [comment]
There is still a ton of test failures due to a change to the error messages in protobuf JSON parser.

**Comment by yanavlasov** [comment]
Still about 15 tests are failing. Mobile build is broken for unknown reason. @RyanTheOptimist

**Comment by yanavlasov** [comment]
Still outstanding issues:

1. mobile build is broken due to its toolchain not supporting compiler warning that new protobuf enables
2. dynamic metadata formatter of NaN and Inf values is broken
3. wasm_config_test and wasm_http_filter_test are broken due to the `Failed to load Wasm module due to a missing import: env.emscripten_stack_snapshot`. Undiagnosed at this point
4. router_tool_test broken, undiagnosed at this point

File: bazel/BUILD
Comment (phlax) [inline bazel/BUILD:878]: trying to understand how this will work

i think the suggestion i made before - ie `"@system_python//:python_headers"` points to locally installed headers 

i wasnt 100% about the suggestion - i saw it in the rules_python source code and various places, and the following ~works

```console
$ bazel build @system_python//:python_headers
```

altho one concern i had was that the system python in the build image 3.10 whereas we are now using 3.11

fwiw there is some work ongoing in `rules_python` to add hermetic py headers to the toolchain

Comment (phlax) [reply bazel/BUILD:878]: iiuc this wont be used/tested in normal circumstances and im not sure its needed at all any more

Comment (yanavlasov) [reply bazel/BUILD:878]: I'm not 100% sure what target is referencing the `:python_headers` as a dependency. `:python_headers` need to actually provide real header files if there is some C++ code that needs python.h which we do not have. I saw that other projects in similar boat just create this dummy target to satisfy bazel. protobuf was doing this this before for us, but now they have removed it so I just defined it locally.

Comment (phlax) [reply bazel/BUILD:878]: ok - checking the (old) upstream and this is basically what it did also (altho it had some notes about what was needed to make it work)

**Review by phlax — DISMISSED** [review]
lgtm, thanks @yanavlasov 

probs best to get a c++ review also

File: source/common/protobuf/yaml_utility.cc
Comment (RyanTheOptimist) [inline source/common/protobuf/yaml_utility.cc:344]: Is this a copy of the function from the old protobuf version? Can we get a brief comment? It's a bit scary that `dst` here doesn't haven a length so we could potentially walk off the end of the buffer. It also looks like we're copying the buffer twice, once on line 364, then again on line 331. How 'bout if we change the signature to:

```
std::string UTF8CoerceToStructurallyValid(absl::string_view str, const char replace_char) {
```
This would create and return a new string which could ensure we don't overflow the string, and would only copy it once. WDYT?

File: source/extensions/filters/http/jwt_authn/filter_config.cc
Comment (RyanTheOptimist) [inline source/extensions/filters/http/jwt_authn/filter_config.cc:59]: `requires_()` is a bit curious. Is this some sort of proto v23 naming change to avoid collisions or some such?

File: test/common/stream_info/stream_info_impl_test.cc
Comment (RyanTheOptimist) [inline test/common/stream_info/stream_info_impl_test.cc]: Out of curiosity, why does this change?

File: test/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter_test.cc
Comment (RyanTheOptimist) [inline test/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter_test.cc:868]: What does the new ":_" suffix mean?

File: test/extensions/filters/network/mongo_proxy/config_test.cc
Comment (RyanTheOptimist) [inline test/extensions/filters/network/mongo_proxy/config_test.cc:148]: Looks like the old error messages were more informative. Is that problematic or are we just minimizing the regex?

**Review by RyanTheOptimist — COMMENTED** [review]
Wow, this looks great. I'm shocked that this is so straightforward. Well done!

File: source/common/formatter/substitution_formatter.cc
Comment (adisuissa) [inline source/common/formatter/substitution_formatter.cc:1967]: This may be a user-facing change.
Is it required, or can it be done in a followup PR (maybe w/ a runtime flag)?

File: bazel/BUILD
Comment (yanavlasov) [reply bazel/BUILD:882]: I've added comment. Hopefully we will never need to make this target supply real headers.

File: test/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter_test.cc
Comment (TAOXUY) [reply test/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter_test.cc:868]: This is introduced by `MessageUtil::codeEnumToString`. Maybe replace it with `absl::StatusCodeToString(code)`.

https://github.com/envoyproxy/envoy/blob/b4e61389f7db842a7bcee89d0fc080b23b27a9d2/source/common/protobuf/utility.cc#L352

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this!
Overall LGTM. Left one comment on something that I think may be a user-facing change

File: source/common/protobuf/yaml_utility.cc
Comment (yanavlasov) [reply source/common/protobuf/yaml_utility.cc:344]: Sounds good

File: source/extensions/filters/http/jwt_authn/filter_config.cc
Comment (yanavlasov) [reply source/extensions/filters/http/jwt_authn/filter_config.cc:59]: `requires` is a C++20 keyword. If it is used to name a proto field, the proto compiler uses `requires_` in generated C++ classes. PGV needs to have the same change otherwise it tried to validate wrong filed name.

File: test/common/stream_info/stream_info_impl_test.cc
Comment (yanavlasov) [reply test/common/stream_info/stream_info_impl_test.cc]: StreamInfo has a proto member. The size of this member has changed with the new protobuf. Some old sizes can be removed later.

File: test/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter_test.cc
Comment (yanavlasov) [reply test/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter_test.cc:868]: Your guess is as good as mine. This is details of protobuf implementation.

File: test/extensions/filters/network/mongo_proxy/config_test.cc
Comment (yanavlasov) [reply test/extensions/filters/network/mongo_proxy/config_test.cc:148]: The error messages are generated internally by protobuf and designed to be human readable. They change from time to time, which already caused churn in such tests before. I have shortened them to ensure that the error message has the offending field present. I think the rest of the error message is not important for the unit test.

File: source/common/formatter/substitution_formatter.cc
Comment (yanavlasov) [reply source/common/formatter/substitution_formatter.cc:1967]: Not, unless we fork the previous implementation of the protobuf JSON formatter. I have discussed this on the maintainers Slack channel and the feedback was that this is edge case that does not warrant forking and maintaining old protobuf code.

File: source/extensions/filters/http/jwt_authn/filter_config.cc
Comment (RyanTheOptimist) [reply source/extensions/filters/http/jwt_authn/filter_config.cc:59]: Ah! That makes sense. Thanks!

File: source/common/protobuf/yaml_utility.cc
Comment (RyanTheOptimist) [inline source/common/protobuf/yaml_utility.cc:330]: nit: I think this line should probably be removed?

**Review by RyanTheOptimist — DISMISSED** [review]
Thanks for driving this!

File: source/common/formatter/substitution_formatter.cc
Comment (adisuissa) [reply source/common/formatter/substitution_formatter.cc:1967]: SGTM, thanks!

File: source/common/protobuf/yaml_utility.cc
Comment (yanavlasov) [reply source/common/protobuf/yaml_utility.cc]: Done

**Review by RyanTheOptimist — DISMISSED** [review]

**Comment by alyssawilk** [comment]
Looks like this one is stalled.  Did we decide to merge now or wait for release?  If we're waiting let's tag it as such just to keep it off the daily on-call spam, and if not let's merge today to max out production time :-)

**Comment by yanavlasov** [comment]
I think we should just delay it, since release is this week

**Review by phlax — APPROVED** [review]
lgtm, thanks @yanavlasov

**Comment by phlax** [comment]
per offline discussion, landing

**Comment by keith** [comment]
it doesn't look like this pgv patch ever made it upstream, sent it here https://github.com/bufbuild/protoc-gen-validate/pull/1110

**Pull request merged**
